src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,551 @@
+    return hashmap
+ 
+
+def create_entries_in_sourcedata_workspace(
+        root_dir: str,
+        env: str,
+        workspace_id: int,
+        assembly_names: list[str],
+        assembly_name_to_dir: dict[str, str],
+        assembly_name_to_upa: dict[str, str],
+        output_dir: str,
+) -> list[str]:
+    """
+    Create a standard entry in sourcedata/workspace for each assembly.
+    Hardlink to the original assembly file in sourcedata to avoid duplicating the file.
+    Update the uploaded.yaml file in the genome directory with assembly names and upa info.
+
+    root_dir: root directory
+    env: KBase environment
+    workspace_id: target workspace addressed by the permanent ID
+    assembly_names: a list of assembly names newly uploaded to the target workspace
+    assembly_name_to_dir: a dictionary of assembly name to its directory path
+    assembly_name_to_upa: a dictionary of assembly name to its UPA
+    output_dir: output directory to create entries in workspace
+    """
+    upas = list()
+    for assembly_name in assembly_names:
+        try:
+            assembly_dir = assembly_name_to_dir[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name}") from e
+
+        src_file = _get_source_file(assembly_dir, assembly_name)
+        assembly_dir_name = os.path.basename(assembly_dir)
+
+        try:
+            upa = assembly_name_to_upa[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name} from target workspace") from e
+        
+        upa_dir = os.path.join(output_dir, upa)
+        os.makedirs(upa_dir, exist_ok=True)
+        upas.append(upa)
+
+        dest_file = os.path.join(upa_dir, assembly_name)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+        _update_yaml_file(root_dir, env, workspace_id, upa, assembly_dir_name, assembly_name)
+
+    return upas
+
+
+def upload_assemblies_to_workspace(
+        conf: Conf,
+        workspace_name: str,
+        data_dir: str,
+) -> list[str]:
+    """
+    Upload assemblies to workspace and record failed assembly names.
+
+    conf: Conf object
+    workspace_name: a string used as a name for a workspace
+    data_dir: directory of assemblies to upload
+    """
+
+    assembly_files = os.listdir(data_dir)
+    print(f'start uploading {len(assembly_files)} assembly files')
+
+    failed_names = list()
+
+    counter = 1
+    for assembly_name in assembly_files:
+
+        if counter % 5000 == 0:
+            print(f"{round(counter / len(assembly_files), 4) * 100}% finished at {datetime.now()}")
+
+        assembly_path = os.path.join(JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, DATA_DIR, assembly_name)
+        try:
+            _upload_assembly_to_workspace(conf, workspace_name, assembly_path, assembly_name)
+        except Exception as e:
+            print(e)
```

Comment:
print the assembly file name as well for debugging purposes

Reply:
Updated!

Reply:
Can you put the assembly name before the exception? Something like:
```python
            print(f"Failed assembly name: {assembly_name}. Exception:")
            print(e)
    
```

Reply:
Updated!

Reply:
:+1: